src/historywork/BatchWork.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "BatchWork.h"
+#include "catchup/CatchupManager.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+
+namespace stellar
+{
+
+BatchWork::BatchWork(Application& app, WorkParent& parent, std::string name)
+    : Work(app, parent, fmt::format("batch-work-" + name), RETRY_NEVER)
+{
+}
+
+BatchWork::~BatchWork()
+{
+    clearChildren();
+}
+
+void
+BatchWork::onReset()
+{
+    resetIter();
+    clearChildren();
+    size_t nChildren = mApp.getConfig().MAX_CONCURRENT_SUBPROCESSES;
+    while (mChildren.size() < nChildren && hasNext())
+    {
+        yieldMoreWork();
+    }
+}
+
+void
+BatchWork::notify(std::string const& child)
+{
+    auto i = mChildren.find(child);
+    if (i == mChildren.end())
+    {
+        CLOG(ERROR, "Work") << "BatchWork notified by unknown child " << child;
+        return;
+    }
+
+    assert(i->second);
+
+    for (auto childIt = mChildren.begin(); childIt != mChildren.end();)
+    {
+        if (childIt->second->getState() == WORK_SUCCESS)
+        {
+            CLOG(DEBUG, "History") << "Finished child work " << childIt->first;
+            childIt = mChildren.erase(childIt);
+            if (hasNext())
```

Comment:
I see that this is taken almost literary from BatchDownloadWork, but only now I have realized that this may not be proper way of adding more work in 100% cases.

Imagine MAX_CONCURRENT_SUBPROCESSES is 16 and at the onReset call there are only 8 work items to do and after first completed work there are additional 32 ones. With current code it would only be able to execute 8 work items in parallel.

I think it would be nice to extract last few lines from onReset to new method runAsManyJobsAsPossible (but with a better name) and call it from onReset and just after this for loop.

Reply:
I'm not sure the scenario where `there are only 8 work items to do and after first completed work there are additional 32 ones` is possible, could you elaborate? I don't think any children of `BatchWork` should allow for range changes (that would cause additional work to be added)

Reply:
This is more for a general-correctness, as I'm not aware of a current work item that allows for range changes.

It serves two, mostly cosmetical, purposes:
1. to split cleaning up completed work items from launching new ones, for more single-responsibility in the code
2. to have only one place in code that launches new work items, removing repetition

Reply:
Ok, I can add that in case there's any work in the future that would add more work while running.

Reply:
Updated.